examples/decoding/plot_ssd_spatial_filters.py

```diff
@@ -28,144 +23,372 @@
+            mean_noise = psd[:, noise_idx].mean(axis=1)
+            spec_ratio = mean_sig / mean_noise
+        sorter_spec = spec_ratio.argsort()[::-1]
+        return spec_ratio, sorter_spec
+
+    def transform(self, inst):
+        """Estimate epochs sources given the SSD filters.
+
+        Parameters
+        ----------
+        inst : instance of Raw, Epochs or ndarray of shape
+            (n_epochs, n_channels, n_times)
+            The data to be processed. The instance is modified inplace.
+
+        Returns
+        -------
+        X_ssd : instance of Raw, Epochs or np.array
+            The processed data.
+        """
+        if self.filters_ is None:
+            raise RuntimeError('No filters available. Please first call fit')
+        if self.return_filtered:
+            inst, _ = self.filter_data(inst)
+        if isinstance(inst, BaseRaw):
+            data = inst.get_data()
+            X_ssd = np.dot(self.filters_.T, data[self.picks_])
+        else:
+            if isinstance(inst, BaseEpochs):
+                X = inst.get_data()[:, self.picks]
+            elif isinstance(inst, np.ndarray):
+                X = inst
+            else:
+                raise NotImplementedError()
+            self._check_X(X)
+            # project data on source space
+            X_ssd = np.asarray(
+                    [np.dot(self.filters_.T, epoch) for epoch in X])
+        if self.sort_by_spectral_ratio:
+            self.spec_ratio, self.sorter_spec = self.spectral_ratio_ssd(
+                ssd_sources=X_ssd)
+            self.filters_ = self.filters_[:, self.sorter_spec]
+            self.patterns_ = self.patterns_[self.sorter_spec]
+
+            if isinstance(inst, BaseRaw):
+                X_ssd = X_ssd[self.sorter_spec]
+            else:
+                X_ssd = X_ssd[:, self.sorter_spec, :]
+            if self.n_components is None:
+                n_components = self.max_components
+                return X_ssd
+            else:
+                n_components = self.n_components
+                if isinstance(inst, BaseRaw):
+                    X_ssd = X_ssd[:n_components]
+                else:
+                    X_ssd = X_ssd[:, :n_components, :]
+                return X_ssd
+
+    def apply(self, inst):
+        """Remove selected components from the signal.
+
+        This procedure will reconstruct M/EEG signals from which the dynamics
+        described by the excluded components is subtracted
+        (denoised by low-rank factorization).
+        See :footcite:'HaufeEtAl2014' for more information.
+
+        The data is processed in place.
+
+        Parameters
+        ----------
+        inst : instance of Raw, Epochs or ndarray
+             The data to be processed. The instance is modified inplace.
+
+        Returns
+        -------
+        X : instance of Raw, Epochs or ndarray
+            The processed data.
+        """
+        X = np.empty_like(inst)
```

Comment:
is this working if you do Epochs/Raw?

Reply:
And is it actually necessary? Seems you're overwriting it.

Reply:
mhm that's true! I deleted that line now !